onnxruntime/contrib_ops/webgpu/quantization/dp4a_matmul_nbits.cc

```diff
@@ -258,11 +258,120 @@ Status DP4AMatMulNBitsProgram::GenerateShaderCode(ShaderHelper& shader) const {
   return Status::OK();
 }
 
+// tile_N size = 16, workgroup size = 64, scale_A components = 1, b components = 4, output components = 4
+Status DP4AMatMulNBitsSmallMProgram::GenerateShaderCode(ShaderHelper& shader) const {
+  shader.AddInput("input_a", ShaderUsage::UseUniform);
+  shader.AddInput("scales_a", ShaderUsage::UseUniform);
+  shader.AddInput("input_b", ShaderUsage::UseUniform);
+  shader.AddInput("scales_b", ShaderUsage::UseUniform);
+  shader.AddOutput("output", ShaderUsage::UseUniform | ShaderUsage::UseElementTypeAlias);
+
+  shader.AdditionalImplementation() << R"ADDNL_FN(
+    const tile_size = 16u; // tile_size = tile_size_vec * output components
```

Comment:
Please leave comment as to why, // tile_size = tile_size_vec * output components should be true.
May be explaining the approach of this shader as a comment would be good as well.

Reply:
This is the tiling of the output ? leaving that note would help as well.

Reply:
I updated the code to make it more flexible and easy to read.
Now you can adjust the workgroup size and tile size in API side very flexibly and don't need to change any code in shader.
With the latest changes, I tested it on NV RTX 2000 Ada, it's faster (45 tokens/s -> 47 tokens/s) than non-dp4a path for f16 type. Please help check the impact on FP32 only GPUs.